07_errors/jimbotech/sync_mapstore.go

```diff
@@ -0,0 +1,81 @@
+package main
+
+import (
+	"math/rand"
+	"sync"
+)
+
+// SyncMapStore stores puppies threadsafe.
+type SyncMapStore struct {
+	mux sync.Mutex
+	sync.Map
+}
+
+// length is not concorrency safe. As the go doc says:
+// Range does not necessarily correspond to any consistent snapshot of the
+// Map's contents: no key will be visited more than once, but if the value
+// for any key is stored or deleted concurrently, Range may reflect any
+// mapping for that key from any point during the Range call.
+//
+func (s *SyncMapStore) length() int {
+	var length int
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	s.Range(func(key interface{}, value interface{}) bool {
+		length++
+		return true
+	})
+	return length
+}
+
+// CreatePuppy threadsafe adding a puppy to storage
+// but will modify the member ID.
+func (s *SyncMapStore) CreatePuppy(p *Puppy) (int32, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	p.ID = rand.Int31()
```

Comment:
you're evil.

Reply:
As above, the gravity of my wickedness is still a mystery to me ... added uniqueness check ...